src/java/com/puppetlabs/jruby_utils/pool/ReferencePool.java

```diff
@@ -0,0 +1,526 @@
+        }
+    }
+
+    public Set<E> getRegisteredElements() {
+      Set<E> registered = new CopyOnWriteArraySet<E>();
+      if (instance != null) {
+          registered.add(instance);
+      }
+      return registered;
+    }
+
+    private void freePoolLock() {
+        poolLockThread = null;
+        // Need to use 'signalAll' here because there might be multiple
+        // waiters (e.g., multiple borrowers) queued up, waiting for the
+        // pool to be unlocked.
+        poolNotLocked.signalAll();
+        // Borrowers that are woken up when an instance is returned to the
+        // pool and the pool borrowLock is held would then start waiting on a
+        // 'poolNotLocked' signal instead.  Re-signalling 'borrowsAvailable' here
+        // allows any borrowers still waiting on the 'borrowsAvailable' signal to be
+        // reawoken when the pool lock is released, compensating for any
+        // 'borrowsAvailable' signals that might have been essentially ignored from
+        // when the pool lock was held.
+        if (this.currentBorrowCount.get() < this.maxBorrowCount) {
+            borrowsAvailable.signalAll();
+        }
+    }
+
+    /**
+     * Should be called if the pool is no longer empty (or a pill is inserted),
+     * so that threads waiting for pool instances can be woken up
+     */
+    private void signalPoolNotEmpty() {
+        // Could use 'signalAll' here instead of 'signal' but 'signal' is
+        // less expensive in that only one waiter will be woken up.  Can use
+        // signal here because the thread being awoken will be able to borrow
+        // a pool instance and any further waiters will be woken up by
+        // subsequent posts of this signal when instances are added/returned to
+        // the queue.
+        borrowsAvailable.signal();
+        signalIfLockCanProceed();
+    }
+
+    /**
+     * Checks if threads waiting on the pool lock should be woken up.
+     * This will wake them up if either all borrowed references have
+     * been returned to the pool, or if a pill has been inserted
+     */
+    private void signalIfLockCanProceed() {
+        // Could use 'signalAll' here instead of 'signal'.  Doesn't really
+        // matter though in that there will only be one waiter at most which
+        // is active at a time - a caller of lock() that has just acquired
+        // the pool lock but is waiting for the live queue to be completely
+        // filled
+        if ((instance != null && currentBorrowCount.get() == 0) || pill != null) {
+            lockAvailable.signal();
+        }
+    }
+
+    private boolean isPoolLockHeld() {
+        return poolLockThread != null;
+    }
+
+    private boolean isPoolLockHeldByCurrentThread(Thread currentThread) {
+        return poolLockThread == currentThread;
+    }
+
+    private boolean isPoolLockHeldByAnotherThread(Thread currentThread) {
+        return (poolLockThread != null) && (poolLockThread != currentThread);
```

Comment:
curious if it would make sense to use `isPoolLockHeld` and `isPoolLockHeldByCurrentThread` in this method. though maybe it's really just tomato tomato.

Reply:
Yeah we should use those methods. I like names.